### PR TITLE
Refacto error parsing (HfHubHttpError)

### PR DESCRIPTION
**Pull request by Wauplin**
(sorry for long PR, most of it is moving logic around)
Done as part of https://github.com/huggingface/huggingface_hub/issues/2069.

This PR updates how `HfHubHttpError` is formatted. With https://github.com/huggingface/huggingface_hub/issues/2069, all errors are now defined in `src/huggingface_hub/errors.py` with no logic it the module. Before this PR, the logic to format the error message from a server response was split between `HfHubHttpError.__init__` and `hf_raise_for_status`. This should not have being split in the first place (harder to read/investigate)

What does this PR do?
- [x] move everything from `HfHubHttpError.__init__` to `hf_raise_for_status`.  This is the only way to instantiate a `HfHubHttpError` anyway
- [x] move `hf_raise_for_status` definition inside `./utils/_http.py` where it belongs instead of `./utils/_errors.py`
- [x] delete `./utils/_errors.py` module to avoid confusion with `./errors.py` where all exceptions are defined
- [x] (fix or feature?) when an error is returned by the server as a raw payload -not json formatted-, we must add the server message to the error message. **This was not the case before.** It is especially useful when TGI returns a 422 unprocessable entity with explanation as raw text
- [x] keep `request_id` / `server_message` as attributes. I'm pretty sure it's not used by anyone but at least it keeps backward compatibility. 
- [x] adapt tests to pass (and remove occurrences of `self.assertEquals...`)

In the end, sorry for the long diff because 95% of it is simply moving some logic around.

### CODE REVIEW

**Comment by Wauplin** [comment]
As spotted by @osanseviero, this will break an import in `datasets-viewer` test suite @severo: https://github.com/huggingface/dataset-viewer/blob/e8e0edfb3e4b76173675d6637c170441abd42141/e2e/tests/utils.py#L15

You can already fix this by doing `from huggingface_hub.utils import hf_raise_for_status`.

File: tests/test_utils_errors.py [hunk 1]
```diff
@@ -11,117 +11,117 @@
     RepositoryNotFoundError,
     RevisionNotFoundError,
 )
-from huggingface_hub.utils._errors import REPO_API_REGEX, hf_raise_for_status
+from huggingface_hub.utils._http import REPO_API_REGEX, X_REQUEST_ID, _format, hf_raise_for_status
```
Comment (osanseviero) [inline tests/test_utils_errors.py:14]: Should the tests be moved to https://github.com/huggingface/huggingface_hub/blob/main/tests/test_utils_http.py for consistency with the rest of the codebase?

**Review by osanseviero — COMMENTED** [review]
Overall LGTM! Will take a second pass

File: src/huggingface_hub/utils/__init__.py [hunk 2]
```diff
@@ -58,6 +55,7 @@
     configure_http_backend,
     fix_hf_endpoint_in_url,
     get_session,
+    hf_raise_for_status,
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply src/huggingface_hub/utils/__init__.py:58]: cc @severo

[the remark this answers is not in the document]
Comment (severo) [reply src/huggingface_hub/utils/__init__.py:58]: indeed, not a big issue I think. It should trigger CI errors when we update the dependency.

[the remark this answers is not in the document]
Comment (severo) [reply src/huggingface_hub/utils/__init__.py:58]: https://github.com/huggingface/dataset-viewer/pull/3034 as proposed in https://github.com/huggingface/huggingface_hub/pull/2474#issuecomment-2302132315

File: tests/test_utils_errors.py [hunk 3]
```diff
@@ -11,117 +11,117 @@
...
-        response.headers = {"X-Error-Code": "RepoNotFound", "X-Request-Id": 123}
+        response.headers = {"X-Error-Code": "RepoNotFound", X_REQUEST_ID: 123}
         response.status_code = 404
         with self.assertRaisesRegex(RepositoryNotFoundError, "Repository Not Found") as context:
             hf_raise_for_status(response)
 
-        self.assertEqual(context.exception.response.status_code, 404)
-        self.assertIn("Request ID: 123", str(context.exception))
+        assert context.exception.response.status_code == 404
+        assert "Request ID: 123" in str(context.exception)
```
Comment (LysandreJik) [inline tests/test_utils_errors.py:26]: Out of curiosity, why the change from `self.assertEqual`/`self.assertIn` to blank asserts with boolean statements? For consistency with new tests?

**Review by LysandreJik — APPROVED** [review]
Thanks for the refactor @Wauplin!

File: tests/test_utils_errors.py (hunk 3 above)
Comment (Wauplin) [reply tests/test_utils_errors.py:26]: Mostly for consistency yes. I'm only doing it when I'm adding tests next to existing ones (see related [convo on slack](https://huggingface.slack.com/archives/C021H1P1HKR/p1707406613622809?thread_ts=1707404663.451959&cid=C021H1P1HKR) -private-)

**Comment by Wauplin** [comment]
Thanks for the review @LysandreJik! I addressed @osanseviero 's comment and resolved the merge conflicts. Failing CI is unrelated so I'm merging now.
